### PR TITLE
Add more tests to string.IndexOf

### PR DESCRIPTION
**Pull request by jamesqo**
dotnet/coreclr#6434 made a bunch of changes to the char-based `string.IndexOf`; here are the tests for them, which go down pretty much all the codepaths in the new implementation. (It basically employs the same trick as dotnet/coreclr#6125, with the exception that it creates a mask out of `value` beforehand and xors each word with that, zeroing out any char in the word that contains `value`.)

I've tested these cases with it as well, and they all seem to pass.

Note that this should probably not be merged until it is, since later changes to the PR may invalidate some of the comments (i.e. the loop unroll size could be changed, which would alter the codepaths some test cases take).

Additional changes:

  - I've also added some test cases for the existing implementation, which is now only used for 32-bit if that PR gets merged.

  - I had to remove all of the `IndexOf` comparisons that didn't use `StringComparison.Ordinal` (by default `IndexOf` uses `CurrentCulture`), since it appears that

  ```cs
  "_____".IndexOf("\ufffd")
  ```

  is 0.

cc @stephentoub

### CODE REVIEW

**Comment by stephentoub** [comment]
> Note that this should probably not be merged until it is, since later changes to the PR may invalidate some of the comments

**Comment by jamesqo** [comment]
@stephentoub OK, done. Even though the PR this was supposed to be for was closed, I decided to keep the tests specific to the 64-bit version just in case I decide to reattempt a similar trick again. (Some of the new tests also cover paths in the current implementation that I don't think were previously being tested.)

**Comment by jamesqo** [comment]
Test Innerloop Ubuntu14.04 Debug Build and Test
Test Innerloop Windows_NT Debug Build and Test

**Comment by stephentoub** [comment]
> I had to remove all of the IndexOf comparisons that didn't use StringComparison.Ordinal (by default IndexOf uses CurrentCulture), 

Isn't this removing useful coverage?  They were passing before; why remove them rather than than just separating out the InlineData that doesn't pass with those?

**Comment by jamesqo** [comment]
@stephentoub Have re-enabled OrdinalIgnoreCase for all overloads, and also the CurrentCulture comparisons for ASCII data.

**Comment by jamesqo** [comment]
Hmm... this is very strange, it's still failing for the `\0` test case I added on Unix, even after I disabled the CurrentCulture tests for control characters. Either I made a mistake in my logic, or something weird is happening with the ordinal/ordinal-ignore-case comparisons that's not happening on Windows.

The CI stack trace doesn't help much, unfortunately (no line number):

```
11:35:29      System.Tests.StringTests.IndexOf_SingleLetter(s: "Hello", target: '\0', startIndex: 0, count: 5, expected: -1) [FAIL]
11:35:29         Assert.Equal() Failure
11:35:29         Expected: -1
11:35:29         Actual:   0
11:35:29         Stack Trace:
11:35:29               at System.Tests.StringTests.IndexOf_SingleLetter(String s, Char target, Int32 startIndex, Int32 count, Int32 expected)
```

**Comment by stephentoub** [comment]
You might be hitting https://github.com/dotnet/coreclr/issues/2051 or some related issue.

**Comment by stephentoub** [comment]
@jamesqo, are you still working on this?

**Comment by jamesqo** [comment]
@stephentoub I've commented out the null char test for now, with a note pointing to the issue you linked to. This should be OK to merge.

File: src/System.Runtime/tests/System/StringTests.cs [hunk 1]
```diff
@@ -1014,24 +1014,99 @@ public static void Format_Invalid()
...
                 }
                 Assert.Equal(expected, s.IndexOf(target, startIndex));
-                Assert.Equal(expected, s.IndexOf(target.ToString(), startIndex));
+                Assert.Equal(expected, s.IndexOf(target.ToString(), startIndex, StringComparison.Ordinal));
+                Assert.Equal(expected, s.IndexOf(target.ToString(), startIndex, StringComparison.OrdinalIgnoreCase));
+
+                if (safeForCurrentCulture)
+                {
+                    Assert.Equal(expected, s.IndexOf(target.ToString(), startIndex));
+                    Assert.Equal(expected, s.IndexOf(target.ToString(), startIndex, StringComparison.CurrentCulture));
```
Comment (stephentoub) [inline src/System.Runtime/tests/System/StringTests.cs]: Do these pass if the current culture is Turkish?

File: src/System.Runtime/tests/System/StringTests.cs [hunk 2]
```diff
@@ -1014,24 +1014,99 @@ public static void Format_Invalid()
...
-            Assert.Equal(expected, s.IndexOf(target.ToString(), startIndex, count));
-
-            Assert.Equal(expected, s.IndexOf(target.ToString(), startIndex, count, StringComparison.CurrentCulture));
             Assert.Equal(expected, s.IndexOf(target.ToString(), startIndex, count, StringComparison.Ordinal));
             Assert.Equal(expected, s.IndexOf(target.ToString(), startIndex, count, StringComparison.OrdinalIgnoreCase));
+
+            if (safeForCurrentCulture)
+            {
+                Assert.Equal(expected, s.IndexOf(target.ToString(), startIndex, count));
+                Assert.Equal(expected, s.IndexOf(target.ToString(), startIndex, count, StringComparison.CurrentCulture));
```
Comment (stephentoub) [inline src/System.Runtime/tests/System/StringTests.cs]: Same question

File: src/System.Runtime/tests/System/StringTests.cs (hunk 1 above)
Comment (jamesqo) [inline src/System.Runtime/tests/System/StringTests.cs]: @stephentoub I just tested by putting

```cs
CultureInfo.CurrentCulture = new CultureInfo("tr-TR");
```

at the beginning of the method. The answer seems to be yes. (These tests only run if all the chars in the string are ASCII and not control.)

Comment (stephentoub) [inline src/System.Runtime/tests/System/StringTests.cs]: Ok. Thanks for checking.

**Pull request merged**